sharding/notary/notary.go

```diff
@@ -151,7 +205,287 @@ func joinNotaryPool(client mainchain.Client) error {
+	}
+
+	if receipt.Status == uint(0) {
+
+		return errors.New("Transaction was not successful, unable to deregister Notary")
+
+	}
+
+	if dreg, err := hasAccountBeenDeregistered(client); dreg || err != nil {
+		if !dreg {
+
+			return errors.New("Notary unable to be Deregistered Successfully from Pool")
+		}
+
+		return err
+	}
+
+	log.Info(fmt.Sprintf("Notary Deregistered from the pool with hash: %s", tx.Hash().String()))
+	return nil
+
+}
+
+// releaseNotary releases the Notary from the pool by deleting the notary from
+// the registry and transferring back the despoit
+func releaseNotary(client mainchain.Client) error {
+	nreg, err := getNotaryRegistry(client)
+	if err != nil {
+		return err
+	}
+	if !nreg.Deposited {
+		return errors.New("Account has not deposited in the Notary Pool")
+	}
+
+	if nreg.DeregisteredPeriod == big.NewInt(0) {
+
+		return errors.New("Notary is still registered to the pool")
+
+	}
+
+	if lockup, err := isLockUpOver(client); !lockup || err != nil {
+		if !lockup {
+			return errors.New("Lockup period is not over")
+		}
+		return err
+	}
+
+	txOps, err := client.CreateTXOpts(nil)
+	if err != nil {
+		return fmt.Errorf("unable to create txOpts: %v", err)
+	}
+	tx, err := client.SMCTransactor().ReleaseNotary(txOps)
+
+	if err != nil {
+		return fmt.Errorf("Unable to Release Notary: %v", err)
+	}
+
+	receipt, err := client.TransactionReceipt(tx.Hash())
+	if err != nil {
+		return err
+	}
+
+	if receipt.Status == uint(0) {
+
+		return errors.New("Transaction was not successful, unable to release Notary")
+
+	}
+
+	nreg, err = getNotaryRegistry(client)
+
+	if err != nil {
+
+		return err
+
+	}
+
+	if nreg.Deposited || nreg.Balance != big.NewInt(0) || nreg.DeregisteredPeriod != big.NewInt(0) || nreg.PoolIndex != big.NewInt(0) {
+
+		return errors.New("Notary unable to be released from the pool")
+
+	}
+
+	log.Info(fmt.Sprintf("Notary with address: %s released from pool", client.Account().Address.String()))
+
+	return nil
+
+}
+
+// submitVote votes for a collation on the shard
+func submitVote(shard sharding.Shard, client mainchain.Client, headerHash *common.Hash) error {
+
+	shardID := shard.ShardID()
+	// checks if the shardID is valid
+	if shardID == big.NewInt(0) || shardID.Int64() > sharding.ShardCount {
+		return fmt.Errorf("ShardId is invalid, it has to be between %d and %d, instead it is %v", 0, sharding.ShardCount, shardID)
+	}
+
+	currentblock, err := client.ChainReader().BlockByNumber(context.Background(), nil)
+	if err != nil {
+		return fmt.Errorf("Unable to retrieve last block: %v", err)
+	}
+
+	period := big.NewInt(0).Div(currentblock.Number(), big.NewInt(sharding.PeriodLength))
+
+	collPeriod, err := client.SMCCaller().LastSubmittedCollation(&bind.CallOpts{}, shardID)
+
+	if err != nil {
+		return fmt.Errorf("Unable to get period from last submitted collation: %v", err)
+	}
+
+	// Checks if the current period is valid in order to vote for the collation on the shard
+	if period.Int64() != collPeriod.Int64() {
```

Comment:
I don't get why this check is for. This check is dangerous imho, what if a period gets skipped and we'll never be able to submit vote again

Reply:
+1

Reply:
I was following this condition in the SMC: 
```solidity
        require(_period == lastSubmittedCollation[_shardId]);
```
But on a second look, it looks like this should be instead 
```solidity
        require(_period > lastSubmittedCollation[_shardId]);
```
right?